pkg/admission/certwriter/certwriter.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certwriter
+
+import (
+	"bytes"
+	"fmt"
+	"net/url"
+
+	admissionregistrationv1beta1 "k8s.io/api/admissionregistration/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"crypto/tls"
+
+	"sigs.k8s.io/controller-runtime/pkg/admission/certgenerator"
+)
+
+const (
+	// CACertName is the name of the CA certificate
+	CACertName = "ca-cert.pem"
+	// ServerKeyName is the name of the server private key
+	ServerKeyName = "key.pem"
+	// ServerCertName is the name of the serving certificate
+	ServerCertName = "cert.pem"
+)
+
+// CertWriter provides method to handle webhooks.
+type CertWriter interface {
+	// EnsureCert ensures that the webhooks it manages have the right certificates.
+	EnsureCert() error
+}
+
+func handleCommon(webhook *admissionregistrationv1beta1.Webhook, ch certReadWriter) error {
+	webhookName := webhook.Name
+	certs, err := ch.read(webhookName)
+	if apierrors.IsNotFound(err) {
+		certs, err = ch.write(webhookName)
+		switch {
+		case apierrors.IsAlreadyExists(err):
+			certs, err = ch.read(webhookName)
+			if err != nil {
+				return err
+			}
+		case err != nil:
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	// Recreate the cert if it's invalid.
+	if !validCert(certs) {
+		certs, err = ch.overwrite(webhookName)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Ensure the CA bundle in the webhook configuration has the signing CA.
+	caBundle := webhook.ClientConfig.CABundle
+	caCert := certs.CACert
+	if !bytes.Contains(caBundle, caCert) {
+		webhook.ClientConfig.CABundle = append(caBundle, caCert...)
+	}
+	return nil
+}
+
+// certReadWriter provides methods for handling certificates for webhook.
+type certReadWriter interface {
+	// read reads a wehbook name and returns the certs for it.
+	read(webhookName string) (*certgenerator.CertArtifacts, error)
+	// write writes the certs and return the certs it wrote.
+	write(webhookName string) (*certgenerator.CertArtifacts, error)
```

Comment:
does this generate them as well?  where does it get the certs from?

Reply:
Yes.
Each `certreadwriter` instance should be able to generate the certs. Mostly it embeds a `CertGenerator`.